papis/commands/tag.py

```diff
@@ -0,0 +1,194 @@
+  same tag cannot exist more than once in a given document.
+
+- To remove a tag use the `--remove` (or `-r`) option:
+
+    .. code:: sh
+
+        papis tag --remove TAG1 QUERY
+
+  This removes TAG1 from the document. If it didn't exist before, nothing
+  happens. You can use `--remove` as many times as you like to remove multiple
+  tags.
+
+- Use `--drop` (or `-d`) to remove all tags:
+
+    .. code:: sh
+
+        papis tag --drop --new TAG1 QUERY
+
+  This removes all tags from the document and then adds TAG1. You can of course
+  also use `--drop` by itself to simply remove all tags without adding new
+  ones.
+
+- Use `--all` (or `-a`) with any of the other options to apply the tagging
+  operation to all matching documents:
+
+    .. code:: sh
+
+        papis tag --all --new TAG1 QUERY
+
+  This adds TAG1 to all documents matching the QUERY rather than opening the
+  picker to let you choose one.
+
+"""
+from typing import List, Optional, Tuple
+
+import click
+
+import papis.cli
+import papis.commands
+import papis.importer
+import papis.commands.update
+import papis.commands.doctor
+import papis.config
+import papis.document
+import papis.strings
+import papis.tui.utils
+
+logger = papis.logging.get_logger(__name__)
+
+
+@click.command("tag")
+@click.help_option("--help", "-h")
+@papis.cli.git_option()
+@papis.cli.bool_flag(
+    "-d",
+    "--drop",
+    help="Drop all tags.",
+    default=False,
+)
+@click.option(
+    "-p",
+    "--add",
+    "to_add",
+    help="Add a tag.",
+    multiple=True,
+    type=str,
+)
+@click.option(
+    "-r",
+    "--remove",
+    "to_remove",
+    help="Remove a tag.",
+    multiple=True,
+    type=str,
+)
+@click.option(
+    "-n",
+    "--rename",
+    "to_rename",
+    help="Rename a tag (<OLD-TAG NEW-TAG>).",
+    multiple=True,
+    type=(str, str),
+)
+@papis.cli.bool_flag("-b", "--batch", help="Batch mode, do not prompt or otherwise")
+@papis.cli.doc_folder_option()
+@papis.cli.all_option()
+@papis.cli.sort_option()
+@papis.cli.query_argument()
+def cli(
+    git: bool,
+    drop: bool,
+    to_add: List[str],
+    to_remove: List[str],
+    to_rename: List[Tuple[str, str]],
+    sort_field: Optional[str],
+    sort_reverse: bool,
+    batch: bool,
+    query: str,
+    doc_folder: Tuple[str, ...],
+    _all: bool,
+) -> None:
+    """
+    Add and delete document tags.
+    """
+
+    if not papis.config.getboolean("tag-ignore-warning"):
+        logger.warning(
+            "This command assumes that your tags are a list (rather than a string). "
+            + "If this is not correct, it will not split them correctly and may give"
+            + "unexpected results. Use `papis doctor -t normalize-tags --fix` to "
+            + "clean up. To disable this warning, set `tag-ignore-warning=true` in "
+            + "your configuration."
+        )
+        if not papis.tui.utils.confirm("Are you sure you want to continue?"):
+            return
```

Comment:
I would say just make this a hard error and tell users to go for `papis doctor` or manual fixing if their tags are not lists. Something like
```
for document in documents:
	if not isinstance(document["tags"], list):
		logger.error(...)
		break

	...
```

For what's implemented so far, we can probably make this work for string-based tags as well. However, going forward, if we want to do fancier things with tags using lists should be much nicer (e.g. don't have to worry about separators or other special characters).

Reply:
Oh, that should also be `papis doctor --check key-type --fix`! There's no `normalize-tags` fix right now, but we can introduce one if it looks nicer..

Reply:
The only problem with this is if the user already uses strings for their tags but happens to not have defined any yet for the documents that are being updated. In that case we can't know whether they are using lists or not. Actually, your proposal (if I'm not mistaken) wouldn't fail in that case, but would abort when the user does use lists but happens not to have defined any tags yet for the relevant documents...

Reply:
I was just reading through the comments (it's been a long time!) and I think what @alejandrogallo wrote up in https://github.com/papis/papis/pull/648#issuecomment-1712909809 makes good sense to me too. So, ideally:
* We should only have one way to store tags and lists seem like the better option.
* We should fail early if the tags are not in the right format. Here we could direct the user to `papis doctor` and/or add a `papis tag --fix` shortcut to do the fixing automatically.

What do you think?

Reply:
>  In that case we can't know whether they are using lists or not. 

In that case we should probably fallback on using lists, since that's the supported format.

Reply:
> We should only have one way to store tags and lists seem like the better option.

I absolutely agree.

> We should fail early if the tags are not in the right format. Here we could direct the user to papis doctor and/or add a papis tag --fix shortcut to do the fixing automatically.

The only worry I have is that users who are using string tags might not know that `papis tag` defaults to lists and then they could make their `info.yaml` files inconsistent (and I think avoiding doing surprising things to people's `info.yaml`'s should be a priority as it might break e.g. their scripts). So, if we do just default to lists, we should be very explicit when releasing the new version that there's a breaking change (of sorts) and that from now on only lists will be supported. Is there maybe a way that a user can be warned once when first using `papis tag` or even just the new version of `papis` (I'm saying 'once' because showing a warning every time definitely gets old fast)?

Reply:
Sounds good, I'll modify it like that!